modules/nw-multus-ipam-object.adoc

```diff
@@ -0,0 +1,186 @@
+// Module included in the following assemblies:
+//
+
+// Because the Cluster Network Operator abstracts the configuration for
+// Macvlan, including IPAM configuration, this must be provided as YAML
+// for the Macvlan CNI plug-in only. In the future other Multus plug-ins
+// may be managed the same way by the CNO.
+
+ifeval::["{context}" == "configuring-macvlan"]
+:yaml:
+endif::[]
+ifeval::["{context}" != "configuring-macvlan"]
+:json:
+endif::[]
+
+:type: pass:q[Specify `static` to configure the plug-in to manage IP address \
+assignment. Specify `DHCP` to allow a DHCP server to manage IP \
+address assignment.]
+
+:addresses: pass:q[that define IP addresses to assign to the virtual interface. Both IPv4 and IPv6 IP addresses are supported.]
+
+:address: pass:q[A block of IP addresses specified in CIDR format to assign to Pods on a worker node.]
+
+:address-gateway: pass:q[The default gateway to route egress network traffic to.]
+
+:routes: pass:q[describing routes to configure inside the Pod.]
+
+:route: pass:q[The IP address range in CIDR format.]
+
+:route-gateway: pass:q[The gateway to use to route network traffic to.]
+
+:dns: pass:q[The DNS configuration. Optional.]
+
+:nameservers: pass:q[one or more IP addresses for to send DNS queries to.]
+
+:domain: pass:q[The default domain to append to a hostname. For example, if the \
+domain is set to `example.com`, a DNS lookup query for `example-host` will be \
+rewritten as `example-host.example.com`.]
+
+:search: pass:q[An array of domain names to append to an unqualified hostname, \
+such as `example-host`, during a DNS lookup query.]
+
+[id="nw-multus-ipam-object_{context}"]
+= Configuration for ipam CNI plug-in
+
+The IP address management (IPAM) CNI plug-in manages IP address assignment for
+other CNI plug-ins.
+
+ifdef::json[]
+The following JSON configuration object describes the parameters you can set:
+endif::json[]
+
+ifdef::yaml[]
+The following YAML configuration describes the parameters you can set:
+endif::yaml[]
+
+ifdef::json[]
+.IPAM CNI plug-in JSON configuration object
+[source,json]
+----
```

Comment:
Hey Jason, if I'm reading this correct... Is this intended to be an overview for all IPAM plugins? I think this JSON block might be too broad to include for all IPAM plugins, unfortunately. 

It's awesome for static, indeed! But, for other types... It's not always the same parameters.

For example, for DHCP, you only set type, for example:

```
{
  "type": "dhcp"
}
```

So while all of these parameters are perfect for the static IPAM CNI plugin-in, the DHCP CNI plug-in is not as feature rich, and doesn't have any additional parameters other than the required type field.

Reply:
@dougbtv, I think I understand what you're saying. I hadn't fully thought through the presentation of the configuration, then. I can think of a few possibilities:

1) simply split the discussion into 2 sections, one for type=static and one for type=DHCP. The configuration for the former would include the various permissible parameters. Then configuration for the latter would simply specify type=DHCP and that no further configuration is possible.

2) Specify for each parameter that it is only applicable to type=static. That will probably get redundant and may be less clear an approach than 1).

I'll have to consider how best to proceed.

Reply:
No worries -- I kinda of lean towards #1 to split into two sections, but, definitely defer to you on how it's presented

Reply:
I'd forgotten, I split this out for the YAML configuration, because there you can only provide the staticIPAMConfig if you are doing static assignment.